### PR TITLE
cmds/core/msr

### PR DESCRIPTION
**Pull request by leongross**
Resolving #3080

### CODE REVIEW

**Review by rminnich — APPROVED** [review]
thank you thank you thank you :-)

**Comment by rminnich** [comment]
This seems to fix all my bugs, why a draft :-)

**Comment by leongross** [comment]
Up until now it only fixes the index out-of-range issue. When I build it on my system, the panic in the interface conversion still exists for said command

```
$ ./msr 0 msr 0x3a reg rd 
panic: interface conversion: forth.Cell is string, not msr.CPUs [recovered]
        panic: interface conversion: forth.Cell is string, not msr.CPUs
```

**Comment by rminnich** [comment]
OK, so... this works
./msr MSR_FEATURE_CONFIG
[0-23 0x13c]

and these little tests work
sudo ./msr "'*" "cpu" "0x13c" "reg"

and we can do this?
sudo ./msr "'*" "cpu" "0x13c" "reg" "rd"

no we can not: 
sudo ./msr "'*" "cpu" "0x13c" "reg" "rd"
panic: [read /dev/cpu/0/msr: input/output error read /dev/cpu/1/msr: input/output error read /dev/cpu/10/msr: input/output error read /dev/cpu/11/msr: input/output error read /dev/cpu/12/msr: input/output error read /dev/cpu/13/msr: input/output error read /dev/cpu/14/msr: input/output error read /dev/cpu/15/msr: input/output error read /dev/cpu/16/msr: input/output error read /dev/cpu/17/msr: input/output error read /dev/cpu/18/msr: input/output error read /dev/cpu/19/msr: input/output error read /dev/cpu/2/msr: input/output error read /dev/cpu/20/msr: input/output error read /dev/cpu/21/msr: input/output error read /dev/cpu/22/msr: input/output error read /dev/cpu/23/msr: input/output error read /dev/cpu/3/msr: input/output error read /dev/cpu/4/msr: input/output error read /dev/cpu/5/msr: input/output error read /dev/cpu/6/msr: input/output error read /dev/cpu/7/msr: input/output error read /dev/cpu/8/msr: input/output error read /dev/cpu/9/msr: input/output error] [recovered]
	panic: interface conversion: string is not error: missing method Error

so what's up here? Well, I screwed up error handling, for one thing. And the msr command seems not to exist any more? Yep.
rminnich@sidecore:~/go/src/github.com/u-root/u-root/cmds/core/msr$ sudo ./msr "'*" "cpu" "0x13c" "reg" "rd"
[]
rminnich@sidecore:~/go/src/github.com/u-root/u-root/cmds/core/msr$ git diff
diff --git a/cmds/core/msr/msr_linux.go b/cmds/core/msr/msr_linux.go
index 19a170369..74dd2c6bb 100644
--- a/cmds/core/msr/msr_linux.go
+++ b/cmds/core/msr/msr_linux.go
@@ -153,9 +153,6 @@ func rd(f forth.Forth) {
        forth.Debug("rd: cpus %v, msr %v", c, r)
        data, errs := r.Read(c)
        forth.Debug("data %v errs %v", data, errs)
-       if errs != nil {
-               panic(fmt.Sprintf("%v", errs))
-       }
        f.Push(data)
 }

This goes back to when I did not think the forth package through; I should have had a stack of interface{}, not string, and msr was written to the older standard.

I'll try to get some fixes in for this.

**Comment by rminnich** [comment]
so:
sudo ./msr "'*" "cpu" "0x13c" "reg" "rd"
[read /dev/cpu/0/msr: input/output error read /dev/cpu/1/msr: input/output error read /dev/cpu/10/msr: input/output error read /dev/cpu/11/msr: input/output error read /dev/cpu/12/msr: input/output error read /dev/cpu/13/msr: input/output error read /dev/cpu/14/msr: input/output error read /dev/cpu/15/msr: input/output error read /dev/cpu/16/msr: input/output error read /dev/cpu/17/msr: input/output error read /dev/cpu/18/msr: input/output error read /dev/cpu/19/msr: input/output error read /dev/cpu/2/msr: input/output error read /dev/cpu/20/msr: input/output error read /dev/cpu/21/msr: input/output error read /dev/cpu/22/msr: input/output error read /dev/cpu/23/msr: input/output error read /dev/cpu/3/msr: input/output error read /dev/cpu/4/msr: input/output error read /dev/cpu/5/msr: input/output error read /dev/cpu/6/msr: input/output error read /dev/cpu/7/msr: input/output error read /dev/cpu/8/msr: input/output error read /dev/cpu/9/msr: input/output error]

diff --git a/cmds/core/msr/msr_linux.go b/cmds/core/msr/msr_linux.go
index 19a170369..16c3b0123 100644
--- a/cmds/core/msr/msr_linux.go
+++ b/cmds/core/msr/msr_linux.go
@@ -154,7 +154,8 @@ func rd(f forth.Forth) {
        data, errs := r.Read(c)
        forth.Debug("data %v errs %v", data, errs)
        if errs != nil {
-               panic(fmt.Sprintf("%v", errs))
+               f.Push(errs)
+               return
        }
        f.Push(data)
 }

**Comment by rminnich** [comment]
I suggest you close this and submit a new one with forth removed. I'm the only one who ever used it.

**Review by rminnich — COMMENTED** [review]
but when you get rid of the forth, be sure to save the symbolic register names.

File: cmds/core/msr/msr_linux.go [hunk 1]
```diff
@@ -149,10 +158,15 @@ func u64(f forth.Forth) {
 
 func rd(f forth.Forth) {
 	r := f.Pop().(msr.MSR)
-	c := f.Pop().(msr.CPUs)
+	// c := f.Pop().(msr.CPUs)
```
Comment (Yonita34567) [inline cmds/core/msr/msr_linux.go:161]: https://github.com/LandrAudio/Falcor-HTTP-WS-DataSource/blob/dependabot/npm_and_yarn/eslint-6.5.1/package.json

**Comment by rminnich** [comment]
I think my #3147 can replace this one.

**Comment by rminnich** [comment]
This should all be fixed, so I'm closing this.

**Pull request closed without merging**